receive_buffer.go

```diff
@@ -39,20 +37,19 @@ func newReceiveBuffer(defaultHeader []byte, ack io.Writer, pool BufferPool, wind
 		in:            make(chan []byte, windowSize),
 		ack:           ack,
 		pool:          pool,
+		closed:        make(chan interface{}),
 	}
 }
 
 // submit allows the session to submit a new frame to the receiveBuffer. If the
 // receiveBuffer has been closed, this is a noop.
 func (buf *receiveBuffer) submit(frame []byte) {
-	buf.mx.RLock()
-	closed := buf.closed
-	if closed {
-		buf.mx.RUnlock()
+	select {
+	case <-buf.closed:
 		return
+	case buf.in <- frame:
+		// okay
 	}
-	buf.in <- frame
```

Comment:
This was the bug. If for some reason we weren't reading from `buf.in`, this could hang and block `buf.mx`, which in turn would block `buf.close`. Reading from `buf.in` can block if `buf.in` is full and we're blocked on sending an ACK. That's not a high percentage thing, which would explain why this doesn't manifest more severely and more quickly, but over time we might be seeing these building up.

The solution is to use a channel rather than a Writer for the acks. Under the covers, that's what we were using anyway, so this just makes the interface a bit more precise and gives us the opportunity to interrupt on close by using a `select`.

Reply:
I think this can be simplified.  [Here's a PR to demonstrate](https://github.com/getlantern/lampshade/pull/50/files).  As far as I can tell, it is logically equivalent, though I could be missing something.

Reply:
Hehe, I actually tried that along the way, but the problem is that it can cause use to lose some of the buffered data. Consider this sequence:

1. Buffer
2. Read from buffer
3. Buffer
4. Close
5. Attempt to read from buffer

On step 5, we want to get what was buffered in step 3 but since we've already closed, it's possible that we won't (`select` is non-deterministic if multiple cases are satisfiable).

So, I had to go back to using a lock, and the only way I could make it work was by adding a timeout.